src/shared/openssl-util.c

```diff
@@ -109,6 +109,126 @@ int rsa_pkey_to_suitable_key_size(
         return 0;
 }
 
+int hkdf_sha512(const uint8_t *in, const size_t in_len,  size_t out_len, uint8_t *ret_out) {
+
+        _cleanup_(EVP_PKEY_CTX_freep) EVP_PKEY_CTX *ctx = NULL;
+        int r = 0;
+
+        assert_se(in);
+        assert_se(in_len);
+        assert_se(out_len);
+
+        ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_HKDF, NULL);
+        if (!ctx)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to allocate pkey context");
+
+        r = EVP_PKEY_derive_init(ctx);
+        if (r != 1)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to initialize pkey derive context");
+
+        r = EVP_PKEY_CTX_hkdf_mode(ctx, EVP_PKEY_HKDEF_MODE_EXTRACT_AND_EXPAND);
+        if (r != 1)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to set hkdf mode");
+
+        r = EVP_PKEY_CTX_set_hkdf_md(ctx, EVP_sha512());
+        if (r != 1)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to set hkdf message digest");
+
+        r = EVP_PKEY_CTX_set1_hkdf_key(ctx, in, in_len);
+        if (r != 1)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to set hkdf key");
+
+        r = EVP_PKEY_derive(ctx, ret_out, &out_len);
+        if (r != 1)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to derive hkdf secret");
+
+        return 0;
+}
+
+int pkey_generate_ec_key(int nid, EVP_PKEY **ret_ppkey) {
+
+        assert_se(ret_ppkey);
+
+        int r = 0;
+        _cleanup_(EVP_PKEY_CTX_freep) EVP_PKEY_CTX *pctx = NULL;
+        _cleanup_(EVP_PKEY_CTX_freep) EVP_PKEY_CTX *kctx = NULL;
+        _cleanup_(EVP_PKEY_freep) EVP_PKEY *params = NULL;
+
+        pctx = EVP_PKEY_CTX_new_id(EVP_PKEY_EC, NULL);
+        if (pctx == NULL)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to allocate pkey context");
+
+        r = EVP_PKEY_paramgen_init((pctx));
+        if (r != 1)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to initialze pkey parameters context");
+
+        r = EVP_PKEY_CTX_set_ec_paramgen_curve_nid(pctx, nid);
+        if (r != 1)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to set pkey context ec curve nid");
+
+        r = EVP_PKEY_paramgen(pctx, &params);
+        if (r != 1)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to generate pkey parameters");
+
+        kctx = EVP_PKEY_CTX_new(params, NULL);
+        if (!kctx)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to allocate pkey context");
+
+        r = EVP_PKEY_keygen_init(kctx);
+        if (r != 1)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to initialize pkey keygen context");
+
+        *ret_ppkey = NULL;
```

Comment:
we generally follow the rule that our functions should initialize all return parameters on success, and non on failure. Here you are initializing ret_ppkey before the last call that can still fail, thus there's a codepath where we'd touch the return param on failure.

thus please, store the result in a local variable first, and only once you know that nothing can fail anymore, copy to the real return param.

Reply:
not adressed yet